Container SSA: warn if no smartproxy/state role

enoodle opened the pull request:
As it is very common to forget to set the SmartProxy role for SmartState Analysis for Container Images I got feedback that a warning if the role is not set is needed. This will prevent the initiation of the job, by disabling the SSA button and will show the user a message with the problem.

![no_smart_proxy_warning](https://cloud.githubusercontent.com/assets/3123328/23256205/3496a478-f9c7-11e6-8423-e94445f58e5b.png)


BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1342790

enoodle commented:
@moolitayer @zgalor Please take a look

zgalor commented:
:+1: LGTM

gshefer commented:
Excellent addition!, I'll add test case for the integration tests once merged.

enoodle commented:
@simon3z

simon3z commented:
@enoodle I think you'll have to sync with @h-kataria @dclarizio and @roliveri to have this on all the SSA pages.

@miq-bot assign enoodle
@miq-bot add_label providers/containers

moolitayer commented on app/controllers/mixins/containers_common_mixin.rb:
```diff
@@ -210,6 +211,14 @@ def process_check_compliance(model, ids)
 
+  def check_smart_roles
+    %w(SmartProxy SmartState).each do |role|
+      unless MiqServer.all.any? { |s| s.has_active_role?(role.downcase) }
```
I'm pretty sure we want to only iterate only the servers in our zone. @gmcculloug ?

enoodle replied:
I could add ` && s.in_active_region?` . waiting on @gmcculloug to ack this.

gmcculloug replied:
Checking the zone sounds like the right way to go, but I have not been involved much with SSA and will defer to @roliveri.

roliveri replied:
@gmcculloug I think that's right. I don't recall, I believe we check for available proxies in zone.

enoodle replied:
@roliveri I am sorry for resurrecting this from the dead ;-)
I was trying to find out but couldn't actually find where we are making this decision. I think it might be in [1] but I am not sure because it seems too general of a function name.
I am asking because I want to answer to BZ [2] with a definitive answer.
I think what they need is to allow Jobs created in the UI zone that have no active roles to be moved to other zones, is that possible?

[1] https://github.com/ManageIQ/manageiq/blob/master/app/models/job_proxy_dispatcher.rb#L174
[2] https://bugzilla.redhat.com/show_bug.cgi?id=1459702

moolitayer commented on app/controllers/mixins/containers_common_mixin.rb:
```diff
@@ -210,6 +211,14 @@ def process_check_compliance(model, ids)
+  def check_smart_roles
+    %w(SmartProxy SmartState).each do |role|
+      unless MiqServer.all.any? { |s| s.has_active_role?(role.downcase) }
+        add_flash("There is no server with the #{role} role enables", :warning)
```
s/enables/enabled/

enoodle replied:
:+1:

moolitayer commented:
And I completely agree we need something against this scenario which keeps happening :+1:

simon3z commented:
@miq-bot assign roliveri

dclarizio commented:
@h-kataria can you take a quick look please?  This looks like a great addition.

dclarizio commented:
@enoodle I like this . . . failing test right now though.

h-kataria commented:
@enoodle in my opinion we should not let user press SSA button if SmartProxy role is not enabled. Iit would be better to disable SSA button on list view & summary screen with a hover text that "There is no server with SmaprtProxy role enabled". You can move the whole logic to check for the enabled role in https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/helpers/application_helper/button/vm_instance_template_scan.rb or to an appropriate button helper class.

enoodle commented:
@h-kataria ok, will do. I just pushed something to make the test green but I will rework this logic for the button type.

enoodle commented:
@h-kataria @dclarizio I changed to disable the button when the required roles are not present. Created a new button type for that need (The VmInstanceTemplateScan button now inherits from it and the container image center is using it directly). I also added tests for this new button.

h-kataria commented:
@enoodle does this disable SSA button on list view as well?

enoodle commented:
@h-kataria I changed the button for `container_images_center.rb` too, but apparently if the only button  in a selection is disabled it won't show this button, so the whole "configuration" button is missing if the role is not set.

h-kataria commented:
@enoodle does this PR address issue mentioned here https://bugzilla.redhat.com/show_bug.cgi?id=1342790

enoodle commented:
@h-kataria yes, I was not aware or this BZ when I created this PR.

h-kataria commented:
@enoodle can you add a link to the BZ in your commit comment.

enoodle commented:
rebased on master to try and fix Travis problems.

simon3z commented on app/helpers/application_helper/button/smart_state_scan.rb:
```diff
@@ -0,0 +1,19 @@
+    my_zone = MiqServer.my_server.my_zone
+    SMART_ROLES.each do |role|
+      unless MiqServer.all.any? { |s| s.has_active_role?(role.downcase) && (s.my_zone == my_zone) }
+        @error_message = "There is no server with the #{role} role enabled"
```
@enoodle @dclarizio does this need i18n?

enoodle replied:
I believe so, I will fix it.

simon3z commented on app/helpers/application_helper/button/smart_state_scan.rb:
```diff
@@ -0,0 +1,19 @@
+  def check_smart_roles
+    my_zone = MiqServer.my_server.my_zone
+    SMART_ROLES.each do |role|
+      unless MiqServer.all.any? { |s| s.has_active_role?(role.downcase) && (s.my_zone == my_zone) }
```
@enoodle if this is the only place where you're using `SMART_ROLES` why don't you keep them `downcase`'d in the constant?

enoodle replied:
I also use it in the message below (line 10).

simon3z commented on app/helpers/application_helper/button/smart_state_scan.rb:
```diff
@@ -0,0 +1,19 @@
+class ApplicationHelper::Button::SmartStateScan < ApplicationHelper::Button::Basic
+  needs :@record
+
+  SMART_ROLES = %w(SmartProxy SmartState).freeze
```
@enoodle @roliveri should this constant live somewhere in the SmartState subsystem?

enoodle replied:
We are using this button for the VM instances too so this should be in a general location. Is there a central SSA class?

simon3z replied:
@enoodle who are you asking? I assume you asked @roliveri (although I am sure you're more than qualified to look at the code and find a relevant place on your own).

enoodle replied:
The only place I could think of is in `MiqServer::ServerSmartProxy` but I am also not sure about it.

[1] https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_server/server_smart_proxy.rb#L3

simon3z commented:
Few comments on the PR but overall LGTM :+1:
@dclarizio @h-kataria can you review/merge?

h-kataria commented:
@simon3z @dclarizio overall changes look good to me, not sure if @enoodle is looking into addressing comments from @simon3z to move the constant to somewhere else.

enoodle commented:
I am waiting on https://github.com/ManageIQ/manageiq/issues/14127 to address the constant for the smart roles.

enoodle commented:
@simon3z, @dclarizio PTAL, I fixed @simon3z 's last comments about the smart proxy roles constant location.

h-kataria commented:
looks good.
